text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+    (preferred) or `connection.try_transaction(f)`.
+
+The ``RetryOptions`` signature:
+
+.. code-block:: typescript
+
+    type DelayFn = (attempt: number) => number;
+    enum AttempsOption {
+        All,
+        NetworkError,
+        ConcurrentUpdate,
+        Deadlock,
+    }
+    class RetryOptions {
+        constructor({
+            attempts: number = 3,
+            transactionOptions?: TransactionOptions,
+        });
+        transactionOptions(TransactionOptions): RetryOptions;
+        attempts(
+            which: AttemptsOption,
+            attempts: number,
+            delay_ms?: DelayFn,
+        ): RetryOptions;
+    }
+
+The ``PoolOptions`` object receives ``retryOptions`` field:
+
+.. code-block:: typescript
+
+    interface PoolOptions {
+      // ...
+      retryOptions?: RetryOptions;
+    }
+
+Exceptions API
+``````````````
+
+Error hierarchy is amended by introducing ``TransientError``, ``NetworkError``
+and ``EarlyNetworkError`` with the following relationships:
+
+.. code-block:: typescript
+
+    class TransientError extends TransactionError {}
+    class TransactionDeadlockError extends TransientError {}
+    class TransactionSerializationError extends TransientError {}
+    class NetworkError extends EdgeDBError {}
+    class EarlyNetworkError extends NetworkError {}
+
+All network error within connection should be converted into
+``EarlyNetworkError`` or ``NetworkError``. Former is used in context
+where we catch network error before sending a request.
+
+
+Python API
+----------
+
+For python API plain ``with`` doesn't work any more, so we introduce a loop
+and with block. See example below.
+
+Pool methods for creating a transaction:
+
+.. code-block:: python
+
+   class AsyncIOPool:
+       def retry(
+           options: RetryOptions = None
+       ) -> AsyncIterable[AsyncContextManager[AsyncIOTransaction]]: ...
+       async def raw_transaction(*,
+           isolation: str = None,
+           readonly: bool = None,
+           deferrable: bool = None,
+       ) -> AsyncContextManager[AsyncIOTransaction]: ...
+
+   class AsyncIOConnection:
+       async def raw_transaction(*,
+           isolation: str = None,
+           readonly: bool = None,
+           deferrable: bool = None,
+       ) -> AsyncContextManager[AsyncIOTransaction]: ...
+
+   class Pool:
+       def retry(
+           options: RetryOptions = None
+       ) -> Iterable[ContextManager[Transaction]]: ...
+       def raw_transaction(
+           isolation: str = None,
+           readonly: bool = None,
+           deferrable: bool = None,
+       ) -> ContextManager[Transaction]: ...
+
+   class Connection:
+       def raw_transaction(
+           isolation: str = None,
+           readonly: bool = None,
+           deferrable: bool = None,
+       ) -> ContextManager[Transaction]: ...
+
+Example usage of ``retry`` on async pool:
+
+.. code-block:: python
+
+    async for tx in db.retry():
+      async with tx:
```

Comment:
I'd also add a quick implementation spec of `retry` in Python. Specifically:

1. `retry()` returns an (async-)iterator which has no methods.
2. Every yielded element is a transaction object, strongly referencing the iterator that created it internally.
3. If the code in the `async with` / `with` block succeeds, the transaction object messages its iterator to stop iteration.

Reply:
Added